File: Trabajo_9/CPP/Ejercicios/Ejercicio_1.cpp
#include <iostream>
using namespace std;
#include <math.h>

class Forma{
public:
    virtual void cambiarPropiedades() = 0;
    virtual void calcularArea() const= 0;
    virtual void calcularPerimetro() const = 0;
    virtual void especificaciones() const = 0;
};


class Circulo : public Forma{
private:
    int radio;
    int diametro;

public:
    Circulo(int r = 0){
        this -> radio = r;
        this -> diametro = 2*radio;
    }

    void cambiarPropiedades() override{
        int r;
        cout << "Ingrese el nuevo radio -> ";
        cin >> r;
        this -> radio = r;
        this -> diametro = 2*r;
    }

    void calcularArea() const override{
        cout << "Area: " << 3.14 * (radio * radio) << endl;
    }

    void calcularPerimetro() const override{
        cout << "Perimetro: " << 3.14 * (diametro) << endl;
    }

    void especificaciones() const override{
        cout << "Datos del Circulo: " << endl;
        cout << "Radio -> " << this -> radio << endl;
        cout << "Diametro -> " << this -> diametro << endl;
        this -> calcularPerimetro();
        this -> calcularArea();
    }
};


class Rectangulo: public Forma{
private:
    int base;
    int altura;

public:
    Rectangulo(int b = 0, int h = 0){
        this -> base = b;
        this -> altura = h;
    }

    void cambiarPropiedades() override{
        int b;
        int h;
        cout << "Ingrese la nueva base -> ";
        cin >> b;
        cout << "Ingrese la nueva altura -> ";
        cin >> h;
        this -> base = b;
        this -> altura = h;
    }

    void calcularArea() const override{
        cout << "Area: " << base * altura << endl;
    }

    void calcularPerimetro() const override{
        cout << "Perimetro: " << 2*(base+altura) << endl;
    }

    void especificaciones() const override{
        cout << "Datos del Rectangulo: " << endl;
        cout << "Base -> " << this -> base << endl;
        cout << "Altura -> " << this -> altura << endl;
        this -> calcularPerimetro();
        this -> calcularArea();
    }
};


class Triangulo: public Forma{
private:
    int base;
    int altura;
    int hipotenusa;

public:
    Triangulo(int b = 0, int h = 0){
        this -> base = b;
        this -> altura = h;
        this -> hipotenusa = sqrt((b*b)+(h*h));
    }

    void cambiarPropiedades() override{
        int b;
        int h;
        cout << "Ingrese la nueva base -> ";
        cin >> b;
        cout << "Ingrese la nueva altura -> ";
        cin >> h;
        this -> base = b;
        this -> altura = h;
        this -> hipotenusa = sqrt((b*b)+(h*h));
    }

    void calcularArea() const override{
        cout << "Area: " << (base * altura)/2 << endl;
    }

    void calcularPerimetro() const override{
        cout << "Perimetro: " << (base+altura+hipotenusa) << endl;
    }

    void especificaciones() const override{
        cout << "Datos del Triangulo: " << endl;
        cout << "Base -> " << this -> base << endl;
        cout << "Altura -> " << this -> altura << endl;
        cout << "Hipotenusa -> " << this -> hipotenusa << endl;
        this -> calcularPerimetro();
        this -> calcularArea();
    }
};

class Trapecio: public Forma{
private:
    int baseMayor;
    int baseMenor;
    int altura;
    int lado1;
    int lado2;

public:
    Trapecio(int bMayor = 0, int bMenor = 0, int h = 0, int l1 = 0, int l2 = 0){
        this -> baseMayor = bMayor;
        this -> baseMenor = bMenor;
        this -> altura = h;
        this -> lado1 = l1;
        this -> lado2 = l2;
    }

    void cambiarPropiedades() override{
        int bMayor, bMenor, h, l1, l2;
        cout << "Ingrese la nueva base mayor -> ";
        cin >> bMayor;
        cout << "Ingrese la nueva base menor -> ";
        cin >> bMenor;
        cout << "Ingrese la nueva altura -> ";
        cin >> h;
        cout << "Ingrese el nuevo lado 1 -> ";
        cin >> l1;
        cout << "Ingrese el nuevo lado 2 -> ";
        cin >> l2;
        this -> baseMayor = bMayor;
        this -> baseMenor = bMenor;
        this -> altura = h;
        this -> lado1 = l1;
        this -> lado2 = l2;
    }

    void calcularArea() const override{
        cout << "Area: " << ((baseMayor + baseMenor) * altura) / 2 << endl;
    }

    void calcularPerimetro() const override{
        cout << "Perimetro: " << baseMayor + baseMenor + lado1 + lado2 << endl;
    }

    void especificaciones() const override{
        cout << "Datos del Trapecio: " << endl;
        cout << "Base Mayor -> " << this -> baseMayor << endl;
        cout << "Base Menor -> " << this -> baseMenor << endl;
        cout << "Altura -> " << this -> altura << endl;
        cout << "Lado 1 -> " << this -> lado1 << endl;
        cout << "Lado 2 -> " << this -> lado2 << endl;
        this -> calcularPerimetro();
        this -> calcularArea();
    }
};

void menu_figuras(){
    cout << "Crear: " << endl;
    cout << "0 -> Finalizar Programa" << endl;
    cout << "1 -> Circulo" << endl;
    cout << "2 -> Rectangulo" << endl;
    cout << "3 -> Triangulo" << endl;
    cout << "4 -> Trapecio" << endl;
}

int validacion(int end){
    int num;
    while (true){
        cin >> num;
        for (int i = 0, end){
            if (num == i){
                return num;
            }
        }
    }
}

int main(){
    while (true){
        menu_figuras()
    }





    return 0;
}

File: Trabajo_9/PY/Ejercicios/Ejercicio_1.py
from abc import ABC, abstractmethod

class Forma(ABC):
    @abstractmethod
    def cambiarPropiedades(self):
        pass
    @abstractmethod
    def calcularArea(self):
        pass
    @abstractmethod
    def calcularPerimetro(self):
        pass
    @abstractmethod
    def especificaciones(self):
        pass

class Circulo(Forma):
    def __init__(self, radio):
        self.radio = radio
        self.diametro = 2*radio

    def cambiarPropiedades(self):
        self.radio = int(input("Ingrese nuevo radio -> "))
        self.diametro = self.radio*2

    def calcularArea(self):
        print(f"Area: {3.14*self.radio**2}")

    def calcularPerimetro(self):
        print(f"Perimetro: {3.14*self.diametro}")

    def especificaciones(self):
        print(f"Datos del circulo\n"
              f"Radio: {self.radio}\n"
              f"Diametro: {self.diametro}")
        self.calcularArea()
        self.calcularPerimetro()

class Rectangulo(Forma):
    def __init__(self, base, altura):
        self.base = base
        self.altura = altura

    def cambiarPropiedades(self):
        self.base = int(input("Ingrese nueva base: "))
        self.altura = int(input("Ingrese nueva altura: "))

    def calcularArea(self):
        print(f"Area: {self.altura*self.base}")

    def calcularPerimetro(self):
        print(f"Perimetro: {2*(self.altura+self.base)}")

    def especificaciones(self):
        print(f"Datos del Rectangulo\n"
              f"Base: {self.base}\n"
              f"Altura: {self.altura}")
        self.calcularArea()
        self.calcularPerimetro()

class Triangulo(Forma):
    def __init__(self, base, altura, hipotenusa):
        self.base = base
        self.altura = altura
        self.hipotenusa = hipotenusa

    def cambiarPropiedades(self):
        self.base = int(input("Ingrese nueva base: "))
        self.altura = int(input("Ingrese nueva altura: "))
        self.hipotenusa = int(input("Ingrese nueva hipotenusa"))

    def calcularArea(self):
        print(f"Area: {0.5*(self.altura*self.base)}")

    def calcularPerimetro(self):
        print(f"Perimetro: {self.altura+self.base+self.hipotenusa}")

    def especificaciones(self):
        print(f"Datos del triangulo\n"
              f"Base: {self.base}\n"
              f"Altura: {self.altura}\n"
              f"Hipotenusa; {self.hipotenusa}")
        self.calcularArea()
        self.calcularPerimetro()

class Trapecio(Forma):
    def __init__(self, baseMenor, baseMayor, altura, lado1, lado2):
        self.baseMenor = baseMenor
        self.baseMayor = baseMayor
        self.altura = altura
        self.lado1 = lado1
        self.lado2 = lado2

    def cambiarPropiedades(self):
        self.baseMenor = int(input("Ingrese nueva Base Menor: "))
        self.baseMayor = int(input("Ingrese nueva Base Mayor: "))
        self.altura = int(input("Ingrese nueva altura: "))
        self.lado1 = int(input("Ingrese nuevo lado1: "))
        self.lado2 = int(input("Ingrese nuevo lado2: "))

    def calcularArea(self):
        print(f"Area: {((self.baseMenor + self.baseMayor)*self.altura)/2}")

    def calcularPerimetro(self):
        print(f"Perimetro: {self.lado1+self.lado2+self.baseMayor+self.baseMenor}")

    def especificaciones(self):
        print(f"Datos del Trapecio\n"
              f"Base Mayor: {self.baseMayor}\n"
              f"Base Menor: {self.baseMenor}\n"
              f"Altura: {self.altura}\n"
              f"Lado 1: {self.lado1}\n"
              f"Lado 2: {self.lado2}")
        self.calcularArea()
        self.calcularPerimetro()

def menu():
    print("Crear Figuras\n"
          " 1.Circulo\n"
          " 2.Rectangulo\n"
          " 3.Triangulo\n"
          " 4.Trapecio\n"
          " 0.Salir")


if __name__ == "__main__":
    figuras = []
    while True:
        menu()
        op = int(input("ingrese una Opción: "))
        if op == 1:
            circulo1 = Circulo(1)
            circulo1.cambiarPropiedades()
            figuras.append(circulo1)
            print("Figura Creada")
        elif op == 2:
            rectangulo1 = Rectangulo(5,4)
            rectangulo1.cambiarPropiedades()
            figuras.append(rectangulo1)
            print("Figura Creada")
        elif op == 3:
            triangulo1 = Triangulo(3,4,5)
            figuras.append(triangulo1)
            print("Figura Creada")
        elif op == 4:
            trapecio1 = Trapecio(5,16,4,5,5)
            figuras.append(trapecio1)
            print("Figura Creada")
        elif op == 0:
            break
        else:
            print("Opcion no valida")

    for figura in figuras:
        figura.especificaciones()
        print("\n")
